src/general/OptionsMenu.cs

```diff
@@ -379,142 +444,241 @@ private Viewport.MSAA MSAAIndexToResolution(int index)
         }
     }
 
+    private void CompareSettings()
+    {
+        // Enable the save and reset buttons if the current setting values differ from the saved ones.
+        if (Settings.Instance == savedSettings)
+        {
+            // Settings match
+            resetButton.Disabled = true;
+            saveButton.Disabled = true;
+        }
+        else
+        {
+            // Settings differ
+            resetButton.Disabled = false;
+            saveButton.Disabled = false;
+        }
+    }
+
     /*
-      GUI callbacks
+      GUI Control Callbacks
     */
 
     private void OnBackPressed()
     {
         GUICommon.Instance.PlayButtonPressSound();
 
-        // TODO: only save if something was changed
+        // If any settings have been changed, show a dialogue asking if the changes should be kept or
+        // discarded.
+        if (Settings.Instance != savedSettings)
+        {
+            backConfirmationBox.PopupCenteredMinsize();
+            return;
+        }
 
-        // TODO: ask for saving the settings
+        EmitSignal(nameof(OnOptionsClosed));
+    }
+
+    private void OnResetPressed()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
+
+        // Restore and apply the old saved settings.
+        Settings.Instance.LoadFromObject(savedSettings);
         Settings.ApplyAll();
+        ApplySettingsToControls(Settings.Instance);
+
+        // Buttons should be disabled now.
+        resetButton.Disabled = true;
```

Comment:
I think it would be better to call `CompareSettings` here.

Reply:
For what, clarity? Unless reset is broken somehow it should always be setting the options back to how they were so I was avoiding another call to compare all those properties if it was unnecessary.

Reply:
There's a bunch of duplicate looking button disable logic spread in 3 methods. Making all of them call the "update buttons to needed state" would be much clearer and less error-prone.

Reply:
Yeah that's fine.